src/plugins/vis_type_metric/public/to_ast.ts

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { get } from 'lodash';
+import { getSchemas, SchemaConfig, Vis } from '../../visualizations/public';
+import { buildExpression, buildExpressionFunction } from '../../expressions/public';
+import { MetricVisExpressionFunctionDefinition } from './metric_vis_fn';
+
+const prepareDimension = (params: SchemaConfig) => {
+  const visdimension = buildExpressionFunction('visdimension', { accessor: params.accessor });
+
+  if (params.format) {
+    visdimension.addArgument('format', params.format.id);
+    visdimension.addArgument('formatParams', JSON.stringify(params.format.params));
+  }
+
+  return buildExpression([visdimension]);
+};
+
+export const toAST = (vis: Vis, params: any) => {
+  // soon this becomes: const esaggs = vis.data.aggs!.toAST();
+  const esaggs = buildExpressionFunction('esaggs', {
+    index: vis.data.indexPattern!.id,
+    metricsAtAllLevels: vis.isHierarchical(),
+    partialRows: vis.type.requiresPartialRows || vis.params.showPartialRows || false,
+    aggConfigs: JSON.stringify(vis.data.aggs!.aggs),
+  });
+
+  const schemas = getSchemas(vis, params);
+
+  const {
+    percentageMode,
+    useRanges,
+    colorSchema,
+    metricColorMode,
+    colorsRange,
+    labels,
+    invertColors,
+    style,
+  } = vis.params.metric;
+
+  // fix formatter for percentage mode
+  if (get(vis.params, 'metric.percentageMode') === true) {
+    schemas.metric.forEach((metric: SchemaConfig) => {
+      metric.format = { id: 'percent' };
+    });
+  }
+
+  // @ts-ignore
+  const metricVis = buildExpressionFunction<MetricVisExpressionFunctionDefinition>('metricVis', {
```

Comment:
curious, what does TS not like about the usage with `buildExpressionFunction` here? wondering if it's something fixable in the function builder

Reply:
we need to cleanup the types i think, we need a separation between argumnent types (arguments user is providing) and argument types (arguments as received inside fn function .... with defaults injected) which we currently don't have.

Reply:
ah, yeah so right now this is a tradeoff between losing type safety when using the expression builder, or having a worse DX when writing the expression function.

you wouldn't need to ignore this if the typings for `metric_vis_fn` args reflected the actual function definition configuration (that is, args which are optional are typed as optional, args which are required are typed as required).

in that case this would work, but it would mean inside the function implementation itself you would need to assert that you know something exists:
```
colorRange: args.colorRange!,
percentageMode: args.percentageMode!,
...etc
```

i think there are a couple ways to solve this which we could discuss -- for now i've pushed an update to this PR to switch out the `ts-ignore` with `ts-expect-error` so that we remember to remove it later when we fix this